### PR TITLE
feat(sast): Implement basic SAST runner

### PR DESCRIPTION
**Pull request by achiar99**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

*Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.*

Fixes # (issue)

### Description
*Include a description of what makes it a violation and any relevant external links.*

### Fix
*How does someone fix the issue in code and/or in runtime?*

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/common/bridgecrew/check_type.py [hunk 1]
```diff
@@ -29,7 +29,15 @@ class CheckType:
     SERVERLESS = "serverless"
     TERRAFORM = "terraform"
     TERRAFORM_PLAN = "terraform_plan"
+    SAST = 'sast'
+
+
+class SastType:
```
Comment (tronxd) [inline checkov/common/bridgecrew/check_type.py:35]:
```suggestion
@dataclass
class SastType:
```

File: checkov/common/bridgecrew/licensing.py [hunk 2]
```diff
@@ -17,7 +18,8 @@ class BillingPlan(str, Enum):
 SubscriptionCategoryMapping = {
     CustomerSubscription.IAC: (CodeCategoryType.IAC, CodeCategoryType.SUPPLY_CHAIN),
     CustomerSubscription.SCA: (CodeCategoryType.OPEN_SOURCE, CodeCategoryType.IMAGES),
-    CustomerSubscription.SECRETS: (CodeCategoryType.SECRETS,)
+    CustomerSubscription.SECRETS: (CodeCategoryType.SECRETS,),
+    CustomerSubscription.SAST: (CodeCategoryType.SAST,),
```
Comment (tronxd) [inline checkov/common/bridgecrew/licensing.py:22]:
```suggestion
    CustomerSubscription.SAST: (CodeCategoryType.SAST,)
```

File: checkov/main.py [hunk 3]
```diff
@@ -113,6 +114,7 @@
     circleci_pipelines_runner(),
     azure_pipelines_runner(),
     ansible_runner(),
+    sast_runner(),
```
Comment (tronxd) [inline checkov/main.py:117]:
```suggestion
    sast_runner()
```

File: checkov/runner_filter.py [hunk 4]
```diff
@@ -338,3 +344,19 @@ def from_dict(obj: Dict[str, Any]) -> RunnerFilter:
     def set_suppressed_policies(self, policy_level_suppressions: List[str]) -> None:
         logging.debug(f"Received the following policy-level suppressions, that will be skipped from running: {policy_level_suppressions}")
         self.suppressed_policies = policy_level_suppressions
+
+    @staticmethod
+    def get_sast_languages(frameworks: Optional[List[str]]) -> Optional[List[SastLanguages]]:
```
Comment (tronxd) [inline checkov/runner_filter.py:349]: Can we test this method?

File: checkov/sast/checks/base_registry.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import logging
+import os
+import yaml
+from typing import List, Any, Optional
+from checkov.common.bridgecrew.check_type import CheckType
+from checkov.common.checks.base_check_registry import BaseCheckRegistry
+from checkov.sast.enums import SastLanguages
+
+CHECKS_POSSIBLE_ENDING = [".yaml", ".yml"]
```
Comment (tronxd) [inline checkov/sast/checks/base_registry.py:11]: import `CHECKS_POSSIBLE_ENDING` from `commons.checks_infra.registry`

File: checkov/sast/checks/base_registry.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+class Registry(BaseCheckRegistry):
+    def __init__(self, checks_dir: str) -> None:
+        super().__init__(report_type=CheckType.SAST)
+        self.rules: List[str] = []
+        self.checks_dir = checks_dir
+        self.logger = logging.getLogger(__name__)
+
+    def extract_entity_details(self, entity: dict[str, Any]) -> tuple[str, str, dict[str, Any]]:
+        return '', '', {}
```
Comment (tronxd) [inline checkov/sast/checks/base_registry.py:22]: Add a TODO for future implementation

File: checkov/sast/runner.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+logger = logging.getLogger(__name__)
+
+
+SEMGREP_SEVERITY_TO_CHECKOV_SEVERITY = {
+    RuleSeverity.ERROR: 'HIGH',
+    RuleSeverity.WARNING: 'MEDIUM',
+    RuleSeverity.INFO: 'LOW',
+}
+
+SUPPORT_FILE_EXT = ['py', 'java', 'js']
```
Comment (tronxd) [inline checkov/sast/runner.py:41]: Maybe couple this mapping into `SastLanguages`? language -> extensions

**Review by tronxd — APPROVED** [review]

File: checkov/runner_filter.py (hunk 4 above)
Comment (achiar99) [reply checkov/runner_filter.py:349]: there is a test

File: checkov/sast/checks/base_registry.py (hunk 5 above)
Comment (achiar99) [reply checkov/sast/checks/base_registry.py:11]: Why? this is not related to graph checks

Comment (tronxd) [reply checkov/sast/checks/base_registry.py:11]: They share the check file format, and unless we change the SAST check format it's a clear code duplication..

File: checkov/runner_filter.py [hunk 8]
```diff
@@ -338,3 +344,19 @@ def from_dict(obj: Dict[str, Any]) -> RunnerFilter:
...
+        for framework in frameworks:
+            if framework == CheckType.SAST:
+                for sast_lang in SastLanguages:
+                    langs.add(sast_lang)
+                return list(langs)
+            if not framework.startswith(CheckType.SAST):
+                continue
+            lang = '_'.join(framework.split('_')[1:])
+            langs.add(SastLanguages[lang.upper()])
+        return list(langs)
```
Comment (gruebel) [inline checkov/runner_filter.py:362]: nothing critical, but why not keep it as a `set`?

File: checkov/sast/checks/base_registry.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        if sast_languages:
+            self._load_checks_from_dir(self.checks_dir, sast_languages)
+
+    def load_external_rules(self, dir: str, sast_languages: Optional[List[SastLanguages]]) -> None:
+        if sast_languages:
+            self._load_checks_from_dir(dir, sast_languages)
+
+    def _load_checks_from_dir(self, directory: str, sast_languages: List[SastLanguages]) -> None:
+        dir = os.path.expanduser(directory)
+        self.logger.debug("Loading external checks from {}".format(dir))
```
Comment (gruebel) [inline checkov/sast/checks/base_registry.py:34]:
```suggestion
        self.logger.debug(f"Loading external checks from {dir}")
```

File: checkov/sast/runner.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            current_min_ident = len(item[1]) - len(item[1].lstrip())
+            if current_min_ident < min_ident:
+                min_ident = current_min_ident
+
+        if min_ident == 0:
+            return code_block
+
+        code_block_cut_ident = []
+        for item in code_block:
+            code_block_cut_ident.append((item[0], item[1][min_ident:]))
```
Comment (gruebel) [inline checkov/sast/runner.py:152]: this can also be changed to a list comprehension

File: checkov/sast/runner.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        return code_block_cut_ident
+
+    @staticmethod
+    def _get_generic_ast(language: SastLanguages, target: str) -> Dict[str, Any]:
+        try:
+            core_runner = CoreRunner(jobs=None, engine=EngineType.OSS, timeout=DEFAULT_TIMEOUT, max_memory=0,
+                                     interfile_timeout=0, timeout_threshold=0, optimizations="none", core_opts_str=None)
+            cmd = [SemgrepCore.path()] + ['-json', '-full_token_info', '-dump_ast', target, '-lang', language.value]
```
Comment (gruebel) [inline checkov/sast/runner.py:160]:
```suggestion
            cmd = [SemgrepCore.path(), '-json', '-full_token_info', '-dump_ast', target, '-lang', language.value]
```
why not fusing both together directly?

File: checkov/sast/runner.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            runner = StreamingSemgrepCore(cmd, 1)
+            runner.vfs_map = {}
+            returncode = runner.execute()
+            output_json: Dict[str, Any] = core_runner._extract_core_output([], returncode, " ".join(cmd), runner.stdout, runner.stderr)
+            return output_json
+        except Exception as e:
+            logger.error(f'Cant parse AST for this file: {target}, for {language.value}: {str(e)}')
```
Comment (gruebel) [inline checkov/sast/runner.py:167]:
```suggestion
        except Exception:
            logger.error(f'Cant parse AST for this file: {target}, for {language.value}', exc_info=True)
```

File: mypy.ini [hunk 13]
```diff
@@ -25,3 +25,6 @@ ignore_missing_imports = True
 
 [mypy-checkov.*]
 follow_imports = skip
+
+[mypy-semgrep.*]
+ignore_missing_imports = True
```
Comment (gruebel) [inline mypy.ini:30]: I can see they don't want people to use it as a library, they have type hints, but didn't add the `py.typed` marker to the project

**Review by gruebel — APPROVED** [review]
nice work 🍻 added a few comments, but nothing critical

**Pull request merged**